Moving NodeElement::press into waitFor.

awesomeousPrime opened the pull request:
The pressing of the node element when pressing a button should also be
surrounded by the waitFor method.

There is a scenario that is attempting to scroll to the bottom where a button is and then press it. What's happening is that at times the scroll is triggered, and the button is attempted to be pressed before the browser has gotten to the desired scroll area. This causes an exception to be thrown but since the $button->press() is outside the waitFor lambda, the button pressing is not re-tried.

The pressing of the button should also be included in a waitFor to try and recover from an error when this happens.

Issue: Medology/stdcheck.com#7989

healthlabs-bugman commented:
Testing file changes were not found.
Paging @juanmrad to review the back-end work in this pull request!
Paging @Chekote to review the testing work in this pull request!

Chekote commented:
Is there not a way that we can more explicitly ensure the button is ready to be pressed, rather than brute force pressing it?

juanmrad commented:
Please check your failing test. also just realized there is no testing on this.

awesomeousPrime commented:
I've gone ahead and fixed the test that failed.  

There was no testing added on this since there's already an existing behat test that asserts the pressButton method passes whenever an item is interacted with outside of the viewport.  This change would just prevent that from intermittently failing if the button was interacted with before the scrolling finished which was what happened in the issue referenced: 
Medology/stdcheck.com#7989

awesomeousPrime commented:
I've gone ahead and fixed the failing test.  Originally there was no testing since the issue is to resolve a possible intermittent failure that happened in the issue mentioned in the description.  I've now added a unit test to asserts that failure of the assertion added, doesn't allow the code to try and press the button found.

Chekote commented on tests/Medology/Behat/Mink/FlexibleContext/PressButtonTest.php:
```diff
@@ -0,0 +1,31 @@
+ */
+class PressButtonTest extends FlexibleContextTest
+{
+    public function testFailingToSeeNodeElementIsVissibleInViewportPreventsButtonFromBeingPressed()
```
You have a typo in the word "Visible".

Chekote commented on tests/Medology/Behat/Mink/FlexibleContext/PressButtonTest.php:
```diff
@@ -0,0 +1,31 @@
+
+        self::expectExceptionMessage('test');
+        $flexible_context->method('assertNodeElementVisibleInViewport')
+            ->willThrowException(new ExpectationException('test', $this->sessionMock));
```
Instead of duplicating the expected error message, you can use the mocked exception you created:

```php
        $exception = new ExpectationException('test', $this->sessionMock);
        $this->expectException(get_class($exception));
        $this->expectExceptionMessage($exception->getMessage());
        $flexible_context->method('assertNodeElementVisibleInViewport')->willThrowException($exception);
```

Also, note how I am asserting the exception class as well as the message, and I am calling the methods via `$this` because they are not static methods.

Chekote commented on tests/Medology/Behat/Mink/FlexibleContext/PressButtonTest.php:
```diff
@@ -0,0 +1,31 @@
+/**
+ * @covers \Medology\Behat\Mink\FlexibleContext::pressButton()
+ */
+class PressButtonTest extends FlexibleContextTest
```
We should also add a test to for:

* Button is not pressed when it is disabled
* Button is pressed when it is enabled and visible
* Correct handling of exceptions that can bubble up from each method called (use data providers)

e.g. getAttribute can throw DriverException & UnsupportedDriverActionException

Chekote commented on src/Medology/Behat/Mink/FlexibleContext.php:
```diff
@@ -1600,6 +1602,25 @@ public function nodeIsFullyVisibleInViewport(NodeElement $element)
+    public function assertNodeElementVisibleInViewport(NodeElement $element)
+    {
+        Spinner::waitFor(function () use ($element) {
+            if (!$this->nodeIsVisibleInViewport($element)) {
```
Please add `@throws` for the two exceptions that can bubble up from this call.

Chekote commented on tests/Medology/Behat/Mink/FlexibleContext/PressButtonTest.php:
```diff
@@ -0,0 +1,150 @@
+        $this->expectException(get_class($exception));
+
+        $flexible_context->pressButton('dsfaljklkj');
+    }
```
I feel like this method is a little complicated. I understand that you did this to remove duplication, but I'm not sure that the additional complexity is worth it. What do you think about the patch below? It's actually 6 fewer lines than what you currently have, and has a little less complexity:

```patch
diff --git a/tests/Medology/Behat/Mink/FlexibleContext/PressButtonTest.php b/tests/Medology/Behat/Mink/FlexibleContext/PressButtonTest.php
index 8f653cd..f38295a 100644
--- a/tests/Medology/Behat/Mink/FlexibleContext/PressButtonTest.php
+++ b/tests/Medology/Behat/Mink/FlexibleContext/PressButtonTest.php
@@ -72,77 +72,70 @@ class PressButtonTest extends FlexibleContextTest
     public function dataFlexibleContextExceptions()
     {
         return [
-            [
-                'flexible_context',
-                'scrollToButton',
-                $this->createMock(ExpectationException::class),
-            ],
-            [
-                'flexible_context',
-                'scrollToButton',
-                $this->createMock(UnsupportedDriverActionException::class),
-            ],
-            [
-                'flexible_context',
-                'assertNodeElementVisibleInViewport',
-                $this->createMock(ExpectationException::class),
-            ],
-            [
-                'flexible_context',
-                'assertNodeElementVisibleInViewport',
-                $this->createMock(UnsupportedDriverActionException::class),
-            ],
-            [
-                'flexible_context',
-                'assertNodeElementVisibleInViewport',
-                $this->createMock(Exception::class),
-            ],
-            [
-                'button',
-                'getAttribute',
-                $this->createMock(DriverException::class),
-            ],
-            [
-                'button',
-                'getAttribute',
-                $this->createMock(UnsupportedDriverActionException::class),
-            ],
-            [
-                'button',
-                'press',
-                $this->createMock(UnsupportedDriverActionException::class),
-            ],
-            [
-                'button',
-                'press',
-                $this->createMock(DriverException::class),
-            ],
+            ['scrollToButton',                     $this->createMock(ExpectationException::class)],
+            ['scrollToButton',                     $this->createMock(UnsupportedDriverActionException::class)],
+            ['assertNodeElementVisibleInViewport', $this->createMock(ExpectationException::class)],
+            ['assertNodeElementVisibleInViewport', $this->createMock(UnsupportedDriverActionException::class)],
+            ['assertNodeElementVisibleInViewport', $this->createMock(Exception::class)],
         ];
     }
 
     /**
-     * Asserts that an exception called from used methods bubbles up.
+     * Asserts that an exception called from FlexibleContext methods bubble up.
      *
-     * @param string    $mock      Name of the mock being tested.
+     * @dataProvider dataFlexibleContextExceptions
      * @param string    $method    Name of method called on mock being tested.
      * @param Exception $exception Exception that should be have bubbled up.
-     *
-     * @dataProvider dataFlexibleContextExceptions
      */
-    public function testExceptionsThrownFromFlexibleContextMethodShouldBubbleOut($mock, $method, Exception $exception)
+    public function testExceptionsThrownFromFlexibleContextMethodsShouldBubbleOut($method, Exception $exception)
     {
         $flexible_context = $this->getFlexibleMock();
         $button = $this->createPartialMock(NodeElement::class, ['getAttribute', 'press']);
+        $button->method('getAttribute')->willReturn('enabled');
 
-        if ($mock != 'button' || $method != 'getAttribute') {
-            $button->method('getAttribute')->willReturn('enabled');
+        if ($method != 'scrollToButton') {
+            $flexible_context->method('scrollToButton')->willReturn($button);
         }
 
-        if ($mock != 'flexible_context' || $method != 'scrollToButton') {
-            $flexible_context->method('scrollToButton')->willReturn($button);
+        $flexible_context->method($method)->willThrowException($exception);
+        $this->expectException(get_class($exception));
+
+        $flexible_context->pressButton('dsfaljklkj');
+    }
+
+    /**
+     * Exceptions thrown when calling specified mock, method combination.
+     *
+     * @return array
+     */
+    public function dataButtonExceptions()
+    {
+        return [
+            ['getAttribute', $this->createMock(DriverException::class)],
+            ['getAttribute', $this->createMock(UnsupportedDriverActionException::class)],
+            ['press',        $this->createMock(DriverException::class)],
+            ['press',        $this->createMock(UnsupportedDriverActionException::class)],
+        ];
+    }
+
+    /**
+     * Asserts that an exception called from Button methods bubble up.
+     *
+     * @dataProvider dataButtonExceptions
+     * @param string    $method    Name of method called on mock being tested.
+     * @param Exception $exception Exception that should be have bubbled up.
+     */
+    public function testExceptionsThrownFromButtonMethodsShouldBubbleOut($method, Exception $exception)
+    {
+        $flexible_context = $this->getFlexibleMock();
+        $button = $this->createPartialMock(NodeElement::class, ['getAttribute', 'press']);
+        $flexible_context->method('scrollToButton')->willReturn($button);
+
+        if ($method != 'getAttribute') {
+            $button->method('getAttribute')->willReturn('enabled');
         }
 
-        ${$mock}->method($method)->willThrowException($exception);
+        $button->method($method)->willThrowException($exception);
         $this->expectException(get_class($exception));
 
         $flexible_context->pressButton('dsfaljklkj');

```

awesomeousPrime replied:
lol sure I'm fine with this.  I had it like this at some point to help me when writing the other method to make it easier to see what things were duplicated.
